ckanext/recombinant/auth.py

```diff
@@ -0,0 +1,42 @@
+from ckan.plugins.toolkit import chained_auth_function, h, _
+from ckan.model.group import Group
+from ckanapi import LocalCKAN, ValidationError
+
+
+@chained_auth_function
+def package_update(up_func, context, data_dict):
+    """
+    Recombinant packages and resources should not be
+    editable by users after they have been made.
+
+    Affects update, patch, and delete.
+    """
+    if data_dict.get(u'type') in h.recombinant_get_types():
+        return {'success': False,
+                'msg': _('User %s not authorized to modify Recombinant type: %s') %
+                            (str(context[u'user']), data_dict.get(u'type'))}
+    return up_func(context, data_dict)
+
+
+@chained_auth_function
+def package_create(up_func, context, data_dict):
+    """
+    If the user has permissions to create packages, they should still
+    be allowed to create the Recombinant packages and resources.
+
+    However, users should not be allowed to create multiple packages
+    of the same type for a single organization.
+    """
+    if data_dict.get(u'type') in h.recombinant_get_types():
+        dataset_type = data_dict.get(u'type')
+        owner_org = Group.get(data_dict.get(u'owner_org'))
+        lc = LocalCKAN(username=context['user'])
+        result = lc.action.package_search(
+            q="type:%s AND organization:%s" % (dataset_type, owner_org.name),
+            include_private=True,
+            rows=2)
+        if result['results']:
+            raise ValidationError({'owner_org':
```

Comment:
auth functions shouldn't raise ValidationError, how about returning success:False here instead?

Reply:
otherwise LGTM

Reply:
Hmm, actually we're passing ignore_auth=True on creating these datasets, right? So the auth functions won't be called at all

Reply:
@wardi I don't think we are ignoring auth here? https://github.com/open-data/ckanext-recombinant/blob/feature/prevent-user-api/ckanext/recombinant/views.py#L394

it uses `LocalCKAN` with the current user.

Reply:
@wardi would using `ignore_auth` in the `localCKAN` allow any logged in user to create a PD record? Even if they are not part of that organization?

Reply:
good point. We should do our own `check_access` call first to see if they have permission to create datasets for this org first.

Reply:
@wardi okay so the POST in the `preview_table` view method now has ignore auth for it's `localCKAN`. And then it does a `check_access` for `package_create`. And the recombinant package create auth is now just super simple.